### PR TITLE
[workflow] Update workflow docs

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

Update workflow docs to adopt recent changes. This PR also deprecates `workflow.init()` in the doc because we are no longer depending on it now.

Note: most examples would not work because we did not support `.options` for workflow steps. The workaround of manipulating `.options(_metadata={"workflow.io/options": {xxx}})` is extremely ugly and I would never give it a chance in the doc.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/workflows/concepts.rst [hunk 1]
```diff
@@ -29,24 +29,48 @@ Concepts
...
 
-    output: "Workflow[int]" = add.step(100, one.step())
+    dag = add.bind(100, one.bind())
+
+Steps
+~~~~~
+
+It takes a single line of code to turn a DAG into a workflow. Each node in the DAG becomes a workflow step.
+
+Steps are retried on failure, but once a step finishes successfully it will never be run again.
```
Comment (iycheng) [inline doc/source/workflows/concepts.rst:65]: We don't have the concept of steps anymore. This is confusing. How about this:

```
It takes a single line of code to turn a DAG into a workflow:

.. code-block:: python
    :caption: Turning the DAG into a workflow:
    from ray import workflow
    output: "Workflow[int]" = workflow.create(dag)


When executing the DAG by the workflow engine, remote functions are retried on failure, but once they finish successfully, they will never be run again.
```

**Comment by suquark** [comment]
@iycheng should we wait until the "DAG doc" is ready so we do not need to mention it in our doc? Or we just go with this doc anyway.

**Comment by iycheng** [comment]
Let's don't couple these two. We should go with this doc and later add a link to the ray dag doc.

I had one initial review of the doc. One issue is that there are still steps in the doc. Maybe we should rename them to node since we don't have the steps concept anymore (they are internal concept)

File: doc/source/workflows/basics.rst [hunk 2]
```diff
@@ -1,73 +1,96 @@
...
 
-Here we created the workflow:
+Here is the workflow we created:
 
 .. image:: basic.png
    :width: 500px
    :align: center
 
-Workflow steps behave similarly to Ray tasks. They are declared via the ``@workflow.step`` annotation, and can take in either concrete values or the output of another workflow step as an argument.
+Each node in the original DAG becomes a workflow step.
```
Comment (iycheng) [inline doc/source/workflows/basics.rst:67]:
```suggestion
Each node in the original DAG becomes a workflow task.
```

File: doc/source/workflows/basics.rst [hunk 3]
```diff
@@ -1,73 +1,96 @@
...
+Workflow steps behave similarly to Ray tasks. They are executed in a parallel and distributed way.
```
Comment (iycheng) [inline doc/source/workflows/basics.rst:68]:
```suggestion
Workflow tasks behave similarly to Ray tasks. They are executed in a parallel and distributed way.
```

File: doc/source/workflows/basics.rst [hunk 4]
```diff
@@ -76,13 +99,13 @@ Workflow results can be retrieved with ``workflow.get_output(workflow_id) -> Obj
 We can retrieve the results for individual workflow steps too with *named steps*. A step can be named in two ways:
```
Comment (iycheng) [inline doc/source/workflows/basics.rst:99]:
```suggestion
We can retrieve the results for individual workflow tasks too with *named steps*. A task can be named in two ways:
```

File: doc/source/workflows/basics.rst [hunk 5]
```diff
@@ -76,13 +99,13 @@ Workflow results can be retrieved with ``workflow.get_output(workflow_id) -> Obj
...
 
  1) via ``.options(name="step_name")``
```
Comment (iycheng) [inline doc/source/workflows/basics.rst:101]:
```suggestion
 1) via ``.options(name="task_name")``
```

File: doc/source/workflows/basics.rst [hunk 6]
```diff
@@ -76,13 +99,13 @@ Workflow results can be retrieved with ``workflow.get_output(workflow_id) -> Obj
...
- 2) via decorator ``@workflow.step(name="step_name"``)
+ 2) via decorator ``@workflow.options(name="step_name"``)
```
Comment (iycheng) [inline doc/source/workflows/basics.rst:102]:
```suggestion
 2) via decorator ``@workflow.options(name="task_name"``)
```

File: doc/source/workflows/basics.rst [hunk 7]
```diff
@@ -76,13 +99,13 @@ Workflow results can be retrieved with ``workflow.get_output(workflow_id) -> Obj
...
 
 Once a step is given a name, the result of the step will be retrievable via ``workflow.get_output(workflow_id, name="step_name")``. If the step with the given name hasn't been executed yet, an exception will be thrown. Here are some examples:
```
Comment (iycheng) [inline doc/source/workflows/basics.rst:104]:
```suggestion
Once a task is given a name, the result of the task will be retrievable via ``workflow.get_output(workflow_id, name="task_name")``. If the task with the given name hasn't been executed yet, an exception will be thrown. Here are some examples:
```

Comment (iycheng) [reply doc/source/workflows/basics.rst:104]: Also for the example below, change step -> task

**Comment by iycheng** [comment]
Could you also update this file: https://github.com/ray-project/ray/blob/master/python/ray/workflow/examples/function_chain.py
(Or delete this)

**Comment by iycheng** [comment]
@suquark  we need to go through the doc and make sure we are not using step(s) there.
I feel we can rename workflow step -> workflow task which seems better and easier to understand.

**Comment by iycheng** [comment]
Also the API page (https://ray--24249.org.readthedocs.build/en/24249/workflows/package-ref.html)
Let's make sure we only expose the necessary APIs there. 
I feel all DAG related one should go to DAG page eventually. We only expose Workflow DAG and management related API.

**Comment by suquark** [comment]
@iycheng I keep them unchanged because removing `step` of them are broken changes. There are two cases: actors and examples with `.options` (waiting https://github.com/ray-project/ray/pull/24458 and a follow up PR to enable workflow options). Should I keep changing them? I think these part of doc will be updated later anyway once we have the proper workflow options API enabled.

**Comment by iycheng** [comment]
That sounds good to me.

**Review by fishbone — APPROVED** [review]

**Comment by suquark** [comment]
The CI failures seem not related. I'll merge it first. A later doc PR will update workflow options.

**Pull request merged**